### DISCUSSION TITLE
How can add more Spacing to Whole beat  and Half beat ?

### DISCUSSION THREAD

**Discussion by goodgame365** [CoderLine/alphaTab#1013] [Q&A] [answered]
now alphatab look like this pic
![1](https://user-images.githubusercontent.com/17884664/195585308-2a380786-4797-49de-a8c7-85eedf36cba2.png)

but it view on guitarpro 8 like this pic

![2](https://user-images.githubusercontent.com/17884664/195585418-977416f0-d212-41ff-ad2b-1452ed4b76dd.png)

gtp file :
[test2.zip](https://github.com/CoderLine/alphaTab/files/9776140/test2.zip)

**Comment by Danielku15** [comment, marked as answer]
The spacing algorithm is a rather fixed algorithm with currently only 1 parameter: [`stretchForce`](https://alphatab.net/docs/reference/settings/display/stretchforce#description)

As described in this page we are using a variant of the Gourlay algorithm.

https://d-nb.info/967375975/34

Looking at the given example generally most music notation software would layout like alphaTab, but Guitar Pro seems to switch into a special spacing mode when having multi-track display active. 

**Guitar Pro 8**
![image](https://user-images.githubusercontent.com/674916/195896841-6683c7f4-4f5b-4d89-8d34-34faae382d69.png)

**alphaTab**
![image](https://user-images.githubusercontent.com/674916/195896949-55361a7c-b1ea-4d68-86ea-98ad3a72aa3e.png)

**MuseScore**
![image](https://user-images.githubusercontent.com/674916/195897052-12a5a0aa-d4a2-4456-8b6c-6f38e7077fad.png)

**SoundSlice**
![image](https://user-images.githubusercontent.com/674916/195897534-0e69f4ca-00d0-42b9-aab3-1972d0b84ae4.png)

**Flat.io**
![image](https://user-images.githubusercontent.com/674916/195897787-7d59bd25-fdf5-4ac8-9710-2132047bedf2.png)

**Guitar Pro 8 - Multi Track**
![image](https://user-images.githubusercontent.com/674916/195897955-ff75d832-246c-4b17-a0ec-80247d34e029.png)

**alphaTab - Multi Track**
![image](https://user-images.githubusercontent.com/674916/195899039-9d156ff1-4e0f-4f92-b4f6-7d80a444d057.png)

**MuseScore - Multi Track**
![image](https://user-images.githubusercontent.com/674916/195898971-c5b2accc-1345-457b-8e0f-62704f1f06ad.png)

**SoundSlice - Multi Track**
![image](https://user-images.githubusercontent.com/674916/195899210-e136a08a-9531-4fb7-9405-59a201c8105c.png)

**Flat.io - Multi Track**
![image](https://user-images.githubusercontent.com/674916/195899637-4905cb40-76b2-4f0a-bebf-d1647c6dc7c4.png)

**Reply by goodgame365** [reply]
A little pity
Because I want to add the Numbered musical notation module 
but it  need more spacing on the half beat and whole

I'll think of another way

**Reply by Danielku15** [reply]
What exactly do you mean by "Numbered musical notation module"? If you share some insight, I might give hints on how to achieve it. e.g. if you need the positions of each note to put annotations you can enable [`includeNoteBounds`](https://alphatab.net/docs/reference/settings/core/includenotebounds#description) and the [`BoundsLookup`](https://alphatab.net/docs/reference/scorerenderer/boundslookup#description) would allow you to get the position of all notes. 


I thought once of a feature like MuseScore to have numbers/names in the note heads:  https://musescore.org/en/handbook/3/notehead-schemes There would be various symbols for that. 

If it goes more towards the Chinese numbered music notation, the starting point would be to make an own `BarRenderer` which is registered in the `staveProfiles` of alphaTab. The `TabBarRenderer` would be a good starting point to create a new renderer which has only a single line and displays certain note values as numbers. 

Technically it would be possible to make also extensions to the spacing algorithms. e.g. to choose that the spacing is just equal on all notes ignoring timing aspects, or having a fixed logarithmic sizing ignoring optimizations like in the Gourlay.

**Comment by goodgame365** [comment]
very thanks for  boss  answer

Indeed, It's  Chinese numbered music notation, I created a new BarRenderer myself，its copy TabBarRenderer

in the Chinese numbered music notation

Whole beat  need four  Quarter  beat spacing , Half beat need two  Quarter  beat spacing  

if Half beat + beat.dots  need  three  Quarter  beat spacing  

 

![22222](https://user-images.githubusercontent.com/17884664/196327374-7022108e-0be9-40cf-87ce-5f08d3302213.png)

**Reply by cao919** [reply]
666 兄嘚 NB plus 能给个Demo吗？

**Reply by Danielku15** [reply]
You might want to keep an eye on https://github.com/CoderLine/alphaTab/issues/1505 for official Chinese numbered music support.
